02-peer-protocol.md

```diff
@@ -430,6 +446,515 @@ would create a Denial of Service risk; therefore, forgetting it is recommended
+
+The opening node may include one output with `sats` value
+of zero. This will be used for change, or discarded if its value is
+be below `dust_limit_satoshis`.
+
+Change is calculated as the sum of the opener's `input_info`.`sats`
+minus the estimated funding transaction size times the
+`feerate_per_kw_funding` minus the `funding_satoshis` minus all other
+`output_info`.`satoshis`. If the `change_satoshis` value is negative,
+the difference is subtracted from `funding_satoshis`. In the case where
+the fee entirely consumes the value of the opener's `funding_satoshis`,
+the remainder will be subtracted from the accepter's `funding_satoshis`.
+Either peer may fail the channel open if the resulting fee payment is
+deemed exorbitant.
+
+A more in-depth discussion of fee payment and change handling can be
+found in [BOLT-3, v2 Funding Transaction Fees](03-transactions.md#channel-establishment-v2-funding-transaction-fees).
+
+`output_info`.`script` is the scriptPubKey for the output.
+
+The channel funding output is not exchanged, as it can be derived
+independently.
+
+
+#### The `funding_add_complete` Message
+
+This message signals the conclusion of a peer's funding transaction
+contributions.
+
+1. type: 62 (`funding_add_complete`)
+2. data:
+    * [`32*byte`:`temporary_channel_id`]
+    * [`u16`:`num_inputs`]
+    * [`u16`:`num_outputs`]
+
+
+##### Requirements
+
+Both nodes:
+  - MUST send this message after `accept_channel2` has been sent/received.
+
+The sending node:
+  - MUST ensure that the `num_inputs` corresponds to the total sum of all `num_inputs`
+    sent in all `funding_add_input` messages that originated from them
+  - MUST ensure that the `num_outputs` corresponds to the total sum of all `num_outputs`
+    sent in all `funding_add_output` messages that originated from them
+
+The receiving node:
+  - MUST fail the channel if:
+    - the `num_inputs` does not correspond to the total sum of all `num_inputs`
+      received in all `funding_add_input` messages
+    - the `num_outputs` does not correspond to the total sum of all `num_outputs`
+      received in all `funding_add_output` messages
+    - the total satoshis of the senders inputs is less than their outputs plus
+      the funding_sats, specified earlier
+  - if is the `opener`:
+    - MAY fail the channel if:
+      - the fee cost of the proposed funding transaction is deemed exorbitant.
+  - if has not yet sent a `funding_add_complete`:
+    - MUST send its `funding_add_complete` message.
+  - otherwise:
+    - MUST use the sent and received `input_info` and `output_info`
+      to create the funding transaction, using `max_witness_len`
+      for each `input_info` and `feerate_per_kw_funding` as specified in
+      `open_channel2`.
+    - MUST send `commitment_signed`.
+
+
+##### Rationale
+
+Each node must have a complete set of the transaction inputs and outputs,
+to derive the funding transaction and subsequent commitment signatures.
```

Comment:
Do we assume nVersion=2 and nLocktime=0 ? If yes we may describe a template in BOLT3 (likely we need one to agree for input order)